### ISSUE TITLE
Segmentation fault while starting VPN

### ISSUE THREAD

**Issue by ashgeek** [adrienverge/openfortivpn#695]
I get this as output while trying to start vpn.

```
INFO:   Connected to gateway.
INFO:   Authenticated.
INFO:   Remote gateway has allocated a VPN.
Segmentation fault
```

This is the output of pppd-log
```
using channel 8
Using interface ppp0
Connect: ppp0 <--> /dev/pts/2
sent [LCP ConfReq id=0x1 <mru 1354> <magic 0x4765c3d0>]
rcvd [LCP ConfReq id=0x1 <magic 0x3d600739>]
sent [LCP ConfAck id=0x1 <magic 0x3d600739>]
rcvd [LCP ConfAck id=0x1 <mru 1354> <magic 0x4765c3d0>]
sent [LCP EchoReq id=0x0 magic=0x4765c3d0]
sent [CCP ConfReq id=0x1 <deflate 15> <deflate(old#) 15> <bsd v1 15>]
sent [IPCP ConfReq id=0x1 <compress VJ 0f 01> <addr 0.0.0.0>]
rcvd [IPCP ConfReq id=0x1 <addr 10.10.10.2>]
sent [IPCP ConfNak id=0x1 <addr 192.0.2.1>]
rcvd [LCP EchoRep id=0x0 magic=0x3d600739]
rcvd [CCP ConfReq id=0x1]
sent [CCP ConfAck id=0x1]
rcvd [CCP ConfRej id=0x1 <deflate 15> <deflate(old#) 15> <bsd v1 15>]
sent [CCP ConfReq id=0x2]
rcvd [IPCP ConfRej id=0x1 <compress VJ 0f 01>]
sent [IPCP ConfReq id=0x2 <addr 0.0.0.0>]
rcvd [IPCP ConfReq id=0x2 <addr 10.10.10.2>]
sent [IPCP ConfNak id=0x2 <addr 192.0.2.1>]
rcvd [CCP ConfAck id=0x2]
rcvd [IPCP ConfNak id=0x2 <addr 192.168.26.92>]
sent [IPCP ConfReq id=0x3 <addr 192.168.26.92>]
rcvd [IPCP ConfReq id=0x3 <addr 10.10.10.2>]
sent [IPCP ConfNak id=0x3 <addr 192.0.2.1>]
rcvd [IPCP ConfAck id=0x3 <addr 192.168.26.92>]
Hangup (SIGHUP)
Modem hangup
Connection terminated.
```

**Comment by DimitriPapadopoulos** [comment]
Which version of _openfortivpn_?

**Comment by ashgeek** [comment]
Hi, thanks for the reply.

Version 1.13.3

verbose output
```
INFO:   Remote gateway has allocated a VPN.
DEBUG:  server_addr: 137.132.84.34
DEBUG:  server_port: 443
DEBUG:  gateway_addr: 137.132.84.34
DEBUG:  gateway_port: 443
DEBUG:  Setting cipher list to: HIGH:!aNULL:!kRSA:!PSK:!SRP:!MD5:!RC4
DEBUG:  Gateway certificate validation succeeded.
DEBUG:  Retrieving configuration
DEBUG:  found dns server 137.132.90.2 in xml config
DEBUG:  found dns server 137.132.85.2 in xml config
DEBUG:  Establishing the tunnel
DEBUG:  ppp_path: /usr/sbin/pppd
DEBUG:  Switch to tunneling mode
DEBUG:  Starting IO through the tunnel
DEBUG:  pppd_read thread
DEBUG:  ssl_read thread
DEBUG:  ssl_write thread
DEBUG:  if_config thread
DEBUG:  pppd_write thread
DEBUG:  pppd ---> gateway (16 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  gateway ---> pppd (16 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  pppd ---> gateway (10 bytes)
DEBUG:  pppd ---> gateway (17 bytes)
DEBUG:  pppd ---> gateway (18 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  gateway ---> pppd (10 bytes)
DEBUG:  gateway ---> pppd (6 bytes)
DEBUG:  gateway ---> pppd (17 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  pppd ---> gateway (6 bytes)
DEBUG:  pppd ---> gateway (6 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  gateway ---> pppd (6 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
Segmentation fault

```

**Comment by DimitriPapadopoulos** [comment]
Strange, 1.13.3 is pretty stable, we've never heard of a segfault before. On which exact OS?

**Comment by ashgeek** [comment]
I see. 

Ubuntu 18.04

**Comment by ashgeek** [comment]
Yeah it might not be a problem with openfortivpn itself.

Check this gdb output. I do not understand this too much. Can you point to what might be wrong?

```
INFO:   Remote gateway has allocated a VPN.
DEBUG:  server_addr: 137.132.84.34
DEBUG:  server_port: 443
DEBUG:  gateway_addr: 137.132.84.34
DEBUG:  gateway_port: 443
DEBUG:  Setting cipher list to: HIGH:!aNULL:!kRSA:!PSK:!SRP:!MD5:!RC4
DEBUG:  Gateway certificate validation succeeded.
DEBUG:  Retrieving configuration
DEBUG:  found dns server 137.132.90.2 in xml config
DEBUG:  found dns server 137.132.85.2 in xml config
DEBUG:  Establishing the tunnel
DEBUG:  ppp_path: /usr/sbin/pppd
DEBUG:  Switch to tunneling mode
DEBUG:  Starting IO through the tunnel
[New Thread 0x7ffff5eab700 (LWP 26937)]
DEBUG:  pppd_read thread
[New Thread 0x7ffff56aa700 (LWP 26938)]
[New Thread 0x7ffff4ea9700 (LWP 26939)]
DEBUG:  ssl_read thread
[New Thread 0x7ffff46a8700 (LWP 26940)]
DEBUG:  ssl_write thread
[New Thread 0x7ffff3ea7700 (LWP 26941)]
DEBUG:  if_config thread
DEBUG:  pppd_write thread
DEBUG:  pppd ---> gateway (16 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  gateway ---> pppd (16 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  pppd ---> gateway (10 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  gateway ---> pppd (10 bytes)
DEBUG:  pppd ---> gateway (17 bytes)
DEBUG:  pppd ---> gateway (18 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  gateway ---> pppd (6 bytes)
DEBUG:  gateway ---> pppd (17 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  pppd ---> gateway (6 bytes)
DEBUG:  pppd ---> gateway (6 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  gateway ---> pppd (6 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
[Switching to Thread 0x7ffff4ea9700 (LWP 26939)]

Thread 4 "openfortivpn" hit Breakpoint 1, __strncat_chk (s1=s1@entry=0x7ffff4ea8d70 "[", 
    s2=0x7ffff4ea96d0 "192.168.26.4", n=n@entry=15, s1len=s1len@entry=327) at strncat_chk.c:25
25	strncat_chk.c: No such file or directory.
(gdb) c
Continuing.
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  pppd ---> gateway (12 bytes)

Thread 4 "openfortivpn" hit Breakpoint 1, __strncat_chk (s1=s1@entry=0x7ffff4ea8d70 "[192.168.26.4], ns [", 
    s2=0x7ffff4ea96d0 "137.132.90.2", n=n@entry=15, s1len=s1len@entry=327) at strncat_chk.c:25
25	in strncat_chk.c
(gdb) c
Continuing.
DEBUG:  pppd ---> gateway (12 bytes)

Thread 4 "openfortivpn" hit Breakpoint 1, __strncat_chk (
    s1=s1@entry=0x7ffff4ea8d70 "[192.168.26.4], ns [137.132.90.2, ", s2=0x7ffff4ea96d0 "137.132.85.2", n=n@entry=15, 
    s1len=s1len@entry=327) at strncat_chk.c:25
25	in strncat_chk.c
(gdb) c
Continuing.

Thread 4 "openfortivpn" hit Breakpoint 1, __strncat_chk (
    s1=s1@entry=0x7ffff4ea8d70 "[192.168.26.4], ns [137.132.90.2, 137.132.85.2], ns_suffix [", s2=0x0, 
    n=n@entry=256, s1len=s1len@entry=327) at strncat_chk.c:25
25	in strncat_chk.c
(gdb) c
Continuing.
DEBUG:  pppd ---> gateway (12 bytes)

Thread 4 "openfortivpn" received signal SIGSEGV, Segmentation fault.
__strncat_chk (s1=0x7ffff4ea8dab "[", 
    s1@entry=0x7ffff4ea8d70 "[192.168.26.4], ns [137.132.90.2, 137.132.85.2], ns_suffix [", 
    s2=0x1 <error: Cannot access memory at address 0x1>, n=n@entry=256, s1len=266, s1len@entry=327)
    at strncat_chk.c:50
50	in strncat_chk.c
(gdb) c
Continuing.
[Thread 0x7ffff3ea7700 (LWP 26941) exited]
[Thread 0x7ffff46a8700 (LWP 26940) exited]
[Thread 0x7ffff4ea9700 (LWP 26939) exited]
[Thread 0x7ffff56aa700 (LWP 26938) exited]
[Thread 0x7ffff5eab700 (LWP 26937) exited]

Program terminated with signal SIGSEGV, Segmentation fault.
The program no longer exists.
```

**Comment by ashgeek** [comment]
Also, it happens in the function call io_loop(&tunnel). More specifically after 
src/io.c: Line 657
```
if (pthread_create(&if_config_thread, NULL, if_config, tunnel))
```  
After the if_config thread is created.

```
DEBUG:  Starting IO through the tunnel
1084		io_loop(&tunnel);
(gdb) 
[New Thread 0x7ffff5eab700 (LWP 32650)]
DEBUG:  pppd_read thread
DEBUG:  pppd ---> gateway (16 bytes)
DEBUG:  pppd ---> gateway (16 bytes)
DEBUG:  pppd ---> gateway (16 bytes)
DEBUG:  pppd ---> gateway (16 bytes)
[New Thread 0x7ffff56aa700 (LWP 32651)]
DEBUG:  pppd_write thread
[New Thread 0x7ffff4ea9700 (LWP 32652)]
DEBUG:  ssl_read thread
[New Thread 0x7fffeffff700 (LWP 32653)]
DEBUG:  ssl_write thread
[New Thread 0x7fffef7fe700 (LWP 32654)]
DEBUG:  if_config thread
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  gateway ---> pppd (16 bytes)
DEBUG:  gateway ---> pppd (16 bytes)
DEBUG:  gateway ---> pppd (16 bytes)
DEBUG:  gateway ---> pppd (16 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  pppd ---> gateway (10 bytes)
DEBUG:  pppd ---> gateway (17 bytes)
DEBUG:  gateway ---> pppd (10 bytes)
DEBUG:  pppd ---> gateway (18 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  gateway ---> pppd (6 bytes)
DEBUG:  gateway ---> pppd (17 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  pppd ---> gateway (6 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  pppd ---> gateway (6 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  gateway ---> pppd (6 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  gateway ---> pppd (12 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  pppd ---> gateway (12 bytes)
DEBUG:  gateway ---> pppd (12 bytes)

Thread 4 "openfortivpn" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff4ea9700 (LWP 32652)]
__strncat_chk (s1=0x7ffff4ea8dac "[", 
    s1@entry=0x7ffff4ea8d70 "[192.168.26.23], ns [137.132.90.2, 137.132.85.2], ns_suffix [", 
    s2=0x1 <error: Cannot access memory at address 0x1>, n=n@entry=256, s1len=265, s1len@entry=327)
    at strncat_chk.c:50
50	strncat_chk.c: No such file or directory.
```

**Comment by DimitriPapadopoulos** [comment]
The following:
  ```
  __strncat_chk (s1=0x7ffff4ea8dac "[", 
      s1@entry=0x7ffff4ea8d70 "[192.168.26.23], ns [137.132.90.2, 137.132.85.2], ns_suffix [", 
      s2=0x1 <error: Cannot access memory at address 0x1>, n=n@entry=256, s1len=265, s1len@entry=327)
  ```
shows the problem occurs here:
https://github.com/adrienverge/openfortivpn/blob/2e53c8edad32cd6844c8678abb85f82aaaaa0fca/src/io.c#L501-L502
and that `tunnel->ipv4.dns_suffix` is null. Indeed this is some new code. It had been introduced by a171722afbc0be0355b2ae920a4fb38ccd8c757c on 9 April 2020 at 22:11:57 +0200 and fixed by f234c5f2650d23daeaeac37b632571e86226861c on 19 April 2020 at 14:27:07 +0200.

I suspect you had retrieved the development version from git between those two dates and built _openfortivpn_ yourself. Just get a more recent version out of git and rebuild. The error will most probably disappear.

**Comment by ashgeek** [comment]
Yeah, I have probably downloaded in between those dates you mentioned.

Now it is fixed. Thanks!
